server/pulp/server/async/app.py

```diff
@@ -49,8 +74,54 @@ def initialize_worker(sender, instance, **kwargs):
     """
     initialization.initialize()
 
+    # Delete any potential old state
     tasks._delete_worker(sender, normal_shutdown=True)
 
     # Create a new working directory for worker that is starting now
     common_utils.delete_worker_working_directory(sender)
     common_utils.create_worker_working_directory(sender)
+
+    # Whether this instance has had to wait for availability
+    _cycled = False
+
+    # If the worker is a resource manager, try to acquire the lock, or wait until it
+    # can be acquired
+    if sender.startswith(constants.RESOURCE_MANAGER_WORKER_NAME):
+
+        name = constants.RESOURCE_MANAGER_WORKER_NAME + "@" + platform.node()
+        lock = ResourceManagerLock(name=name)
+
+        while True:
+
+            # Create a worker record so the user can tell that we're running
+            Worker.objects(name=sender).update_one(set__last_heartbeat=datetime.utcnow(),
+                                                   upsert=True)
+            try:
+                lock.save()
+
+                if _cycled:
+                    # Reset back to the default SIGTERM handler
+                    signal.signal(signal.SIGTERM, signal.SIG_DFL)
+                    msg = _("A new instance of pulp_resource_manager '%s' has become "
+                            "active after the previous instance went offline") % sender
+                    _logger.info(msg)
+
+                break
+            except mongoengine.NotUniqueError:
+                if not _cycled:
+                    # Cleanup routine in event of SIGTERM
+                    def sigterm_handler(_signo, _stack_frame):
+                        msg = _("Worker '%s' shutdown" % name)
+                        _logger.info(msg)
+                        tasks._delete_worker(name, normal_shutdown=True)
+                        sys.exit(0)
+
+                    signal.signal(signal.SIGTERM, sigterm_handler)
+
+                    msg = _("An instance of pulp_resource_manager is already active. "
+                            "Halting '%s' until the currently active "
+                            "resource_manager becomes unavailable." % name)
+                    _logger.info(msg)
+                    _cycled = True
```

Comment:
The replacement of the sigterm handler looks good and correct. It's worth considering a slightly different approach that may help readability and testability.

Consider making a context manager or decorator that adds your handler, then replaces the original. Something like what's below would work, although I'd also consider making ``custom_sigterm_handler`` a decorator, and just decorate the ``get_lock`` function. Either way would be fine, and it might just depend on whether you need control about where in the function the replacement happens.

```
@contextlib.contextmanager
def custom_sigterm_handler(name):
    def sigterm_handler(_signo, _stack_frame):
        msg = _("Worker '%s' shutdown" % name)
        _logger.info(msg)
        tasks._delete_worker(name, normal_shutdown=True)
        sys.exit(0)

    signal.signal(signal.SIGTERM, sigterm_handler)

    yield

    signal.signal(signal.SIGTERM, signal.SIG_DFL)


def get_lock():
    # Whether this instance has had to wait for availability
    _cycled = False

    # If the worker is a resource manager, try to acquire the lock, or wait until it
    # can be acquired
    if sender.startswith(constants.RESOURCE_MANAGER_WORKER_NAME):

        name = constants.RESOURCE_MANAGER_WORKER_NAME + "@" + platform.node()
        lock = ResourceManagerLock(name=name)

        with custom_context_manager():
            # I'm going off memory for this, but you can do something like this to know if the save worked
            while lock.pk is None:

                # Create a worker record so the user can tell that we're running
                Worker.objects(name=sender).update_one(set__last_heartbeat=datetime.utcnow(),
                                                       upsert=True)
                try:
                    lock.save()

                    msg = _("A new instance of pulp_resource_manager '%s' has become "
                            "active.") % sender
                    _logger.info(msg)
                except mongoengine.NotUniqueError:
                    msg = _("An instance of pulp_resource_manager is already active. "
                            "Halting '%s' until the currently active "
                            "resource_manager becomes unavailable." % name)
                    _logger.info(msg)

                    time.sleep(constants.CELERY_CHECK_INTERVAL)
```